### PR TITLE
fix(rust,python): address multiple issues caused by implicit casting of `is_in` values to the column dtype being searched

### LINKED ISSUE #11394: inconsistency with .is_in when filtering string digits
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
df = pl.DataFrame({'a': ["1","2"], 'b': [1,2]})
print(df.filter(pl.col('a').is_in([1, 2])))
print(df.filter(pl.col('a').is_in(["1", "2"])))
print(df.filter(pl.col('a').is_in(["1", 2])))

```

### Log output

_No response_

### Issue description

the digits in column "a" are strings. not sure that the filtering should work when using ints.

### Expected behavior

the digits in column "a" are strings. not sure that the filtering should work when using ints.

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.3
Index type:          UInt32
Platform:            macOS-13.5-arm64-arm-64bit
Python:              3.10.5 (main, Nov 29 2022, 16:45:28) [Clang 14.0.0 (clang-1400.0.29.102)]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.2.1
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.6.0
gevent:              <not installed>
matplotlib:          3.8.0
numpy:               1.26.0
pandas:              2.1.1
pyarrow:             13.0.0
pydantic:            2.4.1
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          3.1.5
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #11394.

* The linked issue shows that we were casting `is_in` values to string if the target series was `Utf8`:
  ```python
  from datetime import date
  import polars as pl

  df = pl.DataFrame({
      "a": ["1","2"],
      "b": [3,4],
  })
  df.filter( pl.col("a").is_in([1,2]) )  # << filtering string by int64 🚫
  # ┌─────┬─────┐
  # │ a   ┆ b   │
  # │ --- ┆ --- │
  # │ str ┆ i64 │
  # ╞═════╪═════╡
  # │ 1   ┆ 3   │
  # │ 2   ┆ 4   │
  # └─────┴─────┘
  ```
* This led down a surprisingly deep rabbit-hole, as various _other_ incompatible dtype pairs were being allowed via implicit cast, leading to incorrect or weird/misleading results - for example, the following filters silently cast and drop all results:
  ```python
  df = pl.DataFrame({
      "a": ["1","2"],
      "b": [3,4],
  })
  df.filter( pl.col("a").is_in([date.today()]) )  # << filtering string by date 🚫
  # ┌─────┬─────┐
  # │ a   ┆ b   │
  # │ --- ┆ --- │
  # │ str ┆ i64 │
  # ╞═════╪═════╡
  # └─────┴─────┘

  df.filter( pl.col("b").is_in([date.today()]) )  # << filtering int64 by date 🚫
  # ┌─────┬─────┐
  # │ a   ┆ b   │
  # │ --- ┆ --- │
  # │ str ┆ i64 │
  # ╞═════╪═════╡
  # └─────┴─────┘
  ```
  I've resolved this by raising `InvalidOperationError` where the dtype of the `is_in` values is not a reasonable match for the dtype of the target Series. We should fail with a clear error here; if the user _wants_ to cast, they can do so explicitly without any trouble. 
  
  In the cases above this means we now raise the following exceptions:
  ```
  InvalidOperationError: 
    `is_in` cannot check for Int64 values in Utf8 data
    `is_in` cannot check for Date values in Int64 data
  ```

* Also spotted that we could return false positives on datetime/duration matches if we allowed `is_in` values with a more granular `time_unit` than the target Series (as we don't cast to supertype); this could also lead to incorrect filtering:
  ```python
  # data has millisecond precision
  df = pl.DataFrame({
      "d": [datetime(2022,10,20,10,30,45,123000)],
  }, schema_overrides={"d": pl.Datetime("ms")})
  
  # filter value has microsecond precision and should NOT match!
  df.filter( 
      pl.col("d").is_in([datetime(2022,10,20,10,30,45,123456)])  # (μs > ms) 🚫
  )
  
  # but... :(
  # ┌─────────────────────────┐
  # │ d                       │
  # │ ---                     │
  # │ datetime[ms]            │
  # ╞═════════════════════════╡
  # │ 2022-10-20 10:30:45.123 │
  # └─────────────────────────┘
  ```
  Solved this by validating that the relative `time_unit` values will not result in silent truncation/rounding. The above example now raises the following:
  ```
  InvalidOperationError: 
    `is_in` cannot check for Microseconds precision values in Milliseconds Datetime data
  ```

(Note that we still allow all of the obvious/expected casts between basic numeric dtypes, etc - only clearly mismatched dtypes -as above- will now raise a suitable exception).

### CODE REVIEW

**Comment by alexander-beedie** [comment]
(Couple of tests/edge-cases to fix up; will finish tomorrow)

File: crates/polars-plan/src/logical_plan/optimizer/type_coercion/mod.rs [hunk 1]
```diff
@@ -356,30 +357,69 @@ impl OptimizationRule for TypeCoercionRule {
...
+                        match (lhs_unit, rhs_unit) {
+                            (TimeUnit::Nanoseconds, _) => return Ok(None),
+                            (TimeUnit::Microseconds, TimeUnit::Microseconds | TimeUnit::Milliseconds) => return Ok(None),
+                            (TimeUnit::Milliseconds, TimeUnit::Milliseconds) => return Ok(None),
+                            _ => polars_bail!(InvalidOperation: "`is_in` cannot check for {:?} precision values in {:?} Datetime data", &rhs_unit, &lhs_unit) 
+                        }
+                    },
+                    (DataType::Duration(lhs_unit), DataType::Duration(rhs_unit)) => {
+                        match (lhs_unit, rhs_unit) {
+                            (TimeUnit::Nanoseconds, _) => return Ok(None),
```
Comment (ritchie46) [inline crates/polars-plan/src/logical_plan/optimizer/type_coercion/mod.rs:379]: I think we can make this a little bit more DRY. We can implement `PartialOrd` for `TimeUnit` at the enum defenition? Than we can check `if lhs_unit <= rhs_unit => Ok(None)`.

Comment (alexander-beedie) [reply crates/polars-plan/src/logical_plan/optimizer/type_coercion/mod.rs:379]: Ahhh, `PartialOrd` is what I was looking for... yes, I didn't like being explicit here 😅
Let me go put on my Rust trousers.

Comment (alexander-beedie) [reply crates/polars-plan/src/logical_plan/optimizer/type_coercion/mod.rs:379]: Done ;)

**Comment by ritchie46** [comment]
Nice one thanks!
